# 01 — Bike Sharing Data Audit

**Analyst:** Fatemeh Kamrani  
**Project:** Bike Sharing Demand Analysis  
**Stage:** Raw Data Audit  

## Purpose

Validate the structure, completeness, consistency, and logical integrity of the raw daily and hourly datasets before any cleaning, transformation, or analysis.

## Inputs

- `../../01_data/raw/day.csv`
- `../../01_data/raw/hour.csv`

## Expected Outputs

- Data quality summary
- Issue register
- Data readiness decision
- Hourly gap diagnostics
> The raw datasets will not be modified in this notebook.

In [1]:
from pathlib import Path

import numpy as np
import pandas as pd



## 1. Load Raw Data
This section defines the project paths and loads the original daily and hourly datasets without modification.

In [3]:
#current notebook location
notebook_dir = Path.cwd()

#main project folder 
project_root = notebook_dir.parents[1]

#raw data folder 
raw_data_dir = project_root /'01_data' /'raw'

#raw data files 
day_file = raw_data_dir /'day.csv'
hour_file =  raw_data_dir /'hour.csv'

print('notebook folder:', notebook_dir)
print('project folder:', project_root)
print('raw data folder:', raw_data_dir)

notebook folder: d:\کلاس دیتا\پاسخ تمارین\مصور سازی\fatemehkamrani_bike_sharing_demand_analysis\02_python\notebooks
project folder: d:\کلاس دیتا\پاسخ تمارین\مصور سازی\fatemehkamrani_bike_sharing_demand_analysis
raw data folder: d:\کلاس دیتا\پاسخ تمارین\مصور سازی\fatemehkamrani_bike_sharing_demand_analysis\01_data\raw


In [4]:
print('day.csv exists: ',day_file.exists())
print('hour.csv exists: ',hour_file.exists())

day.csv exists:  True
hour.csv exists:  True


In [5]:
if not day_file.exists():
    raise FileNotFoundError(f'day.csv was not found in:\n{day_file}')

if not hour_file.exists():
    raise FileNotFoundError(f'hour.csv was not found in:\n{hour_file}')

print('both raw files were found successfully.')

both raw files were found successfully.


In [6]:
day = pd.read_csv(day_file)
hour = pd.read_csv(hour_file)
print('datasets loaded successfully')


datasets loaded successfully


In [7]:
print('day object type:', type(day))
print('hour object type:', type(hour))

day object type: <class 'pandas.DataFrame'>
hour object type: <class 'pandas.DataFrame'>


In [8]:
print('day dataset shape:', day.shape)
print('hour dataset shape:', hour.shape)

day dataset shape: (731, 16)
hour dataset shape: (17379, 17)


## 2. Initial Data Inspection
This section reviews the first rows, column names, dimensions, and data types of the daily and hourly datasets.

In [9]:
day.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [10]:
hour.head()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


In [11]:
day.columns

Index(['instant', 'dteday', 'season', 'yr', 'mnth', 'holiday', 'weekday',
       'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed',
       'casual', 'registered', 'cnt'],
      dtype='str')

In [12]:

print ('day columns:')
print(day.columns.tolist())

print ('hour columns:')
print(hour.columns.tolist())

day columns:
['instant', 'dteday', 'season', 'yr', 'mnth', 'holiday', 'weekday', 'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered', 'cnt']
hour columns:
['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered', 'cnt']


In [13]:
print ('day datasets:')
print ('rows:',day.shape[0])
print ('columns:',day.shape[1])


print ('hour datasets:')
print ('rows:',hour.shape[0])
print ('columns:',hour.shape[1])

day datasets:
rows: 731
columns: 16
hour datasets:
rows: 17379
columns: 17


In [14]:
print ( 'day datasets informations')
day.info()

day datasets informations
<class 'pandas.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    str    
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), str(1)
memory usage: 91.5 KB


In [15]:
print ( 'hour datasets informations')
hour.info()

hour datasets informations
<class 'pandas.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     17379 non-null  int64  
 1   dteday      17379 non-null  str    
 2   season      17379 non-null  int64  
 3   yr          17379 non-null  int64  
 4   mnth        17379 non-null  int64  
 5   hr          17379 non-null  int64  
 6   holiday     17379 non-null  int64  
 7   weekday     17379 non-null  int64  
 8   workingday  17379 non-null  int64  
 9   weathersit  17379 non-null  int64  
 10  temp        17379 non-null  float64
 11  atemp       17379 non-null  float64
 12  hum         17379 non-null  float64
 13  windspeed   17379 non-null  float64
 14  casual      17379 non-null  int64  
 15  registered  17379 non-null  int64  
 16  cnt         17379 non-null  int64  
dtypes: float64(4), int64(12), str(1)
memory usage: 2.3 MB


## 3. Missing Values and Duplicate Records
This section checks for missing values and duplicate records in the daily and hourly datasets before any cleaning decisions are made.

In [16]:
day_missing = day.isna().sum()
hour_missing = hour.isna().sum()
print ('missing values in day dataset:')
display(day_missing)
print ('missing values in hour dataset:')
display(hour_missing)


missing values in day dataset:


instant       0
dteday        0
season        0
yr            0
mnth          0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

missing values in hour dataset:


instant       0
dteday        0
season        0
yr            0
mnth          0
hr            0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

In [17]:
print('total missing values in day dataset:', day.isna().sum().sum())
print('total missing values in hour dataset:', hour.isna().sum().sum())

total missing values in day dataset: 0
total missing values in hour dataset: 0


In [18]:
day_duplicate_rows = day.duplicated().sum()
hour_duplicate_rows = hour.duplicated().sum()

print('duplicate rows in day dataset:', day_duplicate_rows)
print('duplicate rows in hour dataset:', hour_duplicate_rows)

duplicate rows in day dataset: 0
duplicate rows in hour dataset: 0


In [19]:
if day_duplicate_rows > 0:
    display(day[day.duplicated(keep=False)])

if hour_duplicate_rows > 0:
    display(hour[hour.duplicated(keep=False)])

In [20]:
day_duplicate_instant = day['instant'].duplicated().sum()
hour_duplicate_instant = hour['instant'].duplicated().sum()

print('duplicate instant values in day dataset:', day_duplicate_instant)
print('duplicate instant values in hour dataset:', hour_duplicate_instant)

duplicate instant values in day dataset: 0
duplicate instant values in hour dataset: 0


In [21]:
day_duplicate_dates = day['dteday'].duplicated().sum()

print('duplicate dates in day dataset:', day_duplicate_dates)

duplicate dates in day dataset: 0


In [22]:
hour_duplicate_date_hour = hour.duplicated(
    subset=['dteday', 'hr']
).sum()

print(
    'duplicate date-hour combinations in hour dataset:',
    hour_duplicate_date_hour
)

duplicate date-hour combinations in hour dataset: 0


In [23]:
missing_duplicate_summary = pd.DataFrame({
    'dataset': ['day', 'hour'],
    'total_missing_values': [
        day.isna().sum().sum(),
        hour.isna().sum().sum()
    ],
    'duplicate_rows': [
        day.duplicated().sum(),
        hour.duplicated().sum()
    ],
    'duplicate_instant': [
        day['instant'].duplicated().sum(),
        hour['instant'].duplicated().sum()
    ]
})

display(missing_duplicate_summary)

,dataset,total_missing_values,duplicate_rows,duplicate_instant
0,day,0,0,0
1,hour,0,0,0


## 4. Categorical Value Validation
This section checks the unique values of categorical variables to identify invalid or unexpected codes.

In [24]:
categorical_columns = [
    'season',
    'yr',
    'mnth',
    'holiday',
    'weekday',
    'workingday',
    'weathersit'
    
]
print (' unique categorical values in day dataset :\n')

for column in categorical_columns :
    unique_values = sorted (day[column].unique())
    print (column,":", unique_values)

 unique categorical values in day dataset :

season : [np.int64(1), np.int64(2), np.int64(3), np.int64(4)]
yr : [np.int64(0), np.int64(1)]
mnth : [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9), np.int64(10), np.int64(11), np.int64(12)]
holiday : [np.int64(0), np.int64(1)]
weekday : [np.int64(0), np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6)]
workingday : [np.int64(0), np.int64(1)]
weathersit : [np.int64(1), np.int64(2), np.int64(3)]


### Daily categorical validation result

All categorical variables in the daily dataset contain expected codes. No invalid categorical values were identified.

In [25]:
hour_categorical_columns = [
    'season',
    'yr',
    'mnth',
    'hr',
    'holiday',
    'weekday',
    'workingday',
    'weathersit'
]

print('unique categorical values in hour dataset:\n')

for column in hour_categorical_columns:
    unique_values = sorted(map(int, hour[column].unique()))
    print(column, ':', unique_values)

unique categorical values in hour dataset:

season : [1, 2, 3, 4]
yr : [0, 1]
mnth : [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
hr : [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]
holiday : [0, 1]
weekday : [0, 1, 2, 3, 4, 5, 6]
workingday : [0, 1]
weathersit : [1, 2, 3, 4]


### Categorical validation result

All categorical variables in both datasets contain expected values. The hourly dataset also includes weather situation code 4, which is a valid category.

No categorical values require correction or removal.

## 5. Count Variable Validation

This section checks whether rental counts are non-negative and verifies that total rentals equal the sum of casual and registered rentals.

In [26]:
count_validation = pd.DataFrame({
    'dataset': ['day', 'hour'],

    'negative_casual': [
        (day['casual'] < 0).sum(),
        (hour['casual'] < 0).sum()
    ],

    'negative_registered': [
        (day['registered'] < 0).sum(),
        (hour['registered'] < 0).sum()
    ],

    'negative_cnt': [
        (day['cnt'] < 0).sum(),
        (hour['cnt'] < 0).sum()
    ],

    'cnt_mismatch': [
        (day['cnt'] != day['casual'] + day['registered']).sum(),
        (hour['cnt'] != hour['casual'] + hour['registered']).sum()
    ]
})

display(count_validation)

,dataset,negative_casual,negative_registered,negative_cnt,cnt_mismatch
0,day,0,0,0,0
1,hour,0,0,0,0


### Count validation result

No negative rental counts were found. In both datasets, total rentals (`cnt`) are equal to the sum of casual and registered rentals in every record.

No correction is required for the count variables.

## 6. Date Validation

This section checks whether the date values are valid and verifies the date range of each dataset.

In [27]:
day_dates = pd.to_datetime(day['dteday'], errors='coerce')
hour_dates = pd.to_datetime(hour['dteday'], errors='coerce')

print('invalid dates in day dataset:', day_dates.isna().sum())
print('invalid dates in hour dataset:', hour_dates.isna().sum())

print('\nday date range:')
print('start:', day_dates.min())
print('end:', day_dates.max())

print('\nhour date range:')
print('start:', hour_dates.min())
print('end:', hour_dates.max())

invalid dates in day dataset: 0
invalid dates in hour dataset: 0

day date range:
start: 2011-01-01 00:00:00
end: 2012-12-31 00:00:00

hour date range:
start: 2011-01-01 00:00:00
end: 2012-12-31 00:00:00


## 7. Calendar Consistency Validation
This section checks whether the recorded year, month, and weekday values are consistent with the date column.

In [28]:
# calculate expected calendar values from the data column

day_expected_year = day_dates.dt.year -2011
hour_expected_year = hour_dates.dt.year -2011

day_expected_month = day_dates.dt.month
hour_expected_month = hour_dates.dt.month


day_expected_weekday = (day_dates.dt.dayofweek + 1) %7
hour_expected_weekday = (hour_dates.dt.dayofweek + 1) %7




In [29]:
calendar_validation = pd.DataFrame({
    'dataset': ['day', 'hour'],

    'year_mismatch': [
        (day['yr'] != day_expected_year).sum(),
        (hour['yr'] != hour_expected_year).sum()
    ],

    'month_mismatch': [
        (day['mnth'] != day_expected_month).sum(),
        (hour['mnth'] != hour_expected_month).sum()
    ],

    'weekday_mismatch': [
        (day['weekday'] != day_expected_weekday).sum(),
        (hour['weekday'] != hour_expected_weekday).sum()
    ]
})

display(calendar_validation)

,dataset,year_mismatch,month_mismatch,weekday_mismatch
0,day,0,0,0
1,hour,0,0,0


### Calendar consistency result

The recorded year, month, and weekday values are fully consistent with the date column in both datasets.

No calendar inconsistencies were identified.

In [30]:
day_expected_workingday = (
    (day['holiday'] == 0) &
    (~day['weekday'].isin([0, 6]))
).astype(int)

hour_expected_workingday = (
    (hour['holiday'] == 0) &
    (~hour['weekday'].isin([0, 6]))
).astype(int)

day_workingday_mismatch = (
    day['workingday'] != day_expected_workingday
).sum()

hour_workingday_mismatch = (
    hour['workingday'] != hour_expected_workingday
).sum()

print(
    'workingday mismatches in day dataset:',
    day_workingday_mismatch
)

print(
    'workingday mismatches in hour dataset:',
    hour_workingday_mismatch
)

workingday mismatches in day dataset: 0
workingday mismatches in hour dataset: 0


### Working-day validation result
The working-day variable is consistent with the holiday and weekday variables in both datasets.
No working-day inconsistencies were identified.

## 8. Hourly Record Completeness
This section checks how many hourly records are available for each date and identifies dates with fewer than 24 hourly observations.


In [31]:
hour_records_per_day = (
    hour.groupby('dteday').size()
    .reset_index(name= 'hour_record_count')
)
display (hour_records_per_day.head())

,dteday,hour_record_count
0,2011-01-01,24
1,2011-01-02,23
2,2011-01-03,22
3,2011-01-04,23
4,2011-01-05,23


In [32]:
complete_days = (
    hour_records_per_day['hour_record_count'] == 24
).sum()

incomplete_days = (
    hour_records_per_day['hour_record_count'] < 24
).sum()

print('dates with 24 hourly records:', complete_days)
print('dates with fewer than 24 hourly records:', incomplete_days)

dates with 24 hourly records: 655
dates with fewer than 24 hourly records: 76


In [33]:
incomplete_hourly_dates = hour_records_per_day[
    hour_records_per_day['hour_record_count'] < 24
].sort_values('hour_record_count')

display(incomplete_hourly_dates.head(20))

,dteday,hour_record_count
667,2012-10-29,1
26,2011-01-27,8
668,2012-10-30,11
17,2011-01-18,12
25,2011-01-26,16
239,2011-08-28,17
238,2011-08-27,18
52,2011-02-22,18
68,2011-03-10,22
58,2011-02-28,22


### Hourly completeness result

The hourly dataset contains 655 complete dates with 24 hourly records and 76 incomplete dates with fewer than 24 records.

This represents structural time gaps rather than cell-level missing values. No missing hourly records will be imputed during the raw data audit.

In [34]:
expected_hour_records = len(hour_records_per_day) * 24
observed_hour_records = len(hour)

missing_hour_records = (
    expected_hour_records - observed_hour_records
)

print('expected hourly records:', expected_hour_records)
print('observed hourly records:', observed_hour_records)
print('missing hourly records:', missing_hour_records)

expected hourly records: 17544
observed hourly records: 17379
missing hourly records: 165


In [35]:
expected_hours = set(range(24))

missing_hours_by_date = (
    hour.groupby('dteday')['hr']
    .apply(
        lambda hours: sorted(
            expected_hours - set(hours)
        )
    )
    .reset_index(name='missing_hours')
)

missing_hours_by_date['missing_hour_count'] = (
    missing_hours_by_date['missing_hours']
    .apply(len)
)

missing_hours_by_date = missing_hours_by_date[
    missing_hours_by_date['missing_hour_count'] > 0
].sort_values(
    'missing_hour_count',
    ascending=False
)

display(missing_hours_by_date.head(20))

,dteday,missing_hours,missing_hour_count
667,2012-10-29,"[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14...",23
26,2011-01-27,"[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...",16
668,2012-10-30,"[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]",13
17,2011-01-18,"[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]",12
25,2011-01-26,"[3, 4, 18, 19, 20, 21, 22, 23]",8
239,2011-08-28,"[0, 1, 2, 3, 4, 5, 6]",7
238,2011-08-27,"[18, 19, 20, 21, 22, 23]",6
52,2011-02-22,"[0, 1, 2, 3, 4, 5]",6
41,2011-02-11,"[3, 4]",2
2,2011-01-03,"[2, 3]",2


### Hourly completeness finding

The hourly dataset contains 17,379 observed records, compared with 17,544 expected date-hour combinations. Therefore, 165 hourly records are absent across 76 dates.

The gaps are structural because the corresponding rows do not exist in the dataset. They are not cell-level missing values. No hourly records will be imputed or removed during the raw data audit.

In [36]:
issue_register = pd.DataFrame({
    'issue_id': ['dq-01'],
    'dataset': ['hour'],
    'issue': ['missing date-hour records'],
    'description': [
        '165 expected date-hour combinations are absent across 76 dates.'
    ],
    'severity': ['medium'],
    'decision': [
        'retain the raw data and evaluate the effect before hourly analysis'
    ]
})

display(issue_register)

,issue_id,dataset,issue,description,severity,decision
0,dq-01,hour,missing date-hour records,165 expected date-hour combinations are absent...,medium,retain the raw data and evaluate the effect be...


## 9. Daily and Hourly Demand Reconciliation

This section aggregates hourly rental counts by date and compares them with the corresponding daily totals.

In [37]:
hourly_daily_totals = (
    hour.groupby('dteday', as_index=False)[
        ['casual', 'registered', 'cnt']
    ]
    .sum()
    .rename(columns={
        'casual': 'hour_casual',
        'registered': 'hour_registered',
        'cnt': 'hour_cnt'
    })
)

daily_totals = (
    day[
        ['dteday', 'casual', 'registered', 'cnt']
    ]
    .rename(columns={
        'casual': 'day_casual',
        'registered': 'day_registered',
        'cnt': 'day_cnt'
    })
)

In [38]:
daily_hourly_check = daily_totals.merge(
    hourly_daily_totals,
    on='dteday',
    how='left'
)

daily_hourly_check['casual_mismatch'] = (
    daily_hourly_check['day_casual'] !=
    daily_hourly_check['hour_casual']
)

daily_hourly_check['registered_mismatch'] = (
    daily_hourly_check['day_registered'] !=
    daily_hourly_check['hour_registered']
)

daily_hourly_check['cnt_mismatch'] = (
    daily_hourly_check['day_cnt'] !=
    daily_hourly_check['hour_cnt']
)

In [39]:
demand_reconciliation_summary = pd.DataFrame({
    'variable': ['casual', 'registered', 'cnt'],
    'mismatch_dates': [
        daily_hourly_check['casual_mismatch'].sum(),
        daily_hourly_check['registered_mismatch'].sum(),
        daily_hourly_check['cnt_mismatch'].sum()
    ]
})

display(demand_reconciliation_summary)

,variable,mismatch_dates
0,casual,0
1,registered,0
2,cnt,0


### Daily-hourly reconciliation result

The aggregated hourly rental counts exactly match the corresponding daily totals for casual, registered, and total rentals.
No demand inconsistencies were identified between the daily and hourly datasets. However, the previously identified structural hourly gaps remain relevant for hour-level analysis.

## 10. Weather Variable Range Validation

This section checks the minimum and maximum values of the normalized weather variables and identifies values outside the expected range of 0 to 1.

In [40]:
weather_columns = [
    'temp',
    'atemp',
    'hum',
    'windspeed'
]
day_weather_range = day[weather_columns].agg(
    ['min', 'max']
).T
hour_weather_range = hour[weather_columns].agg(
    ['min', 'max']
).T
print('weather variable ranges in day dataset:')
display(day_weather_range)
print('weather variable ranges in hour dataset:')
display(hour_weather_range)



weather variable ranges in day dataset:


,min,max
temp,0.059130,0.861667
atemp,0.079070,0.840896
hum,0.000000,0.972500
windspeed,0.022392,0.507463


weather variable ranges in hour dataset:


,min,max
temp,0.02,1.0000
atemp,0.00,1.0000
hum,0.00,1.0000
windspeed,0.00,0.8507


In [41]:
day_weather_out_of_range = (
    (day[weather_columns] < 0) |
    (day[weather_columns] > 1)
).sum()

hour_weather_out_of_range = (
    (hour[weather_columns] < 0) |
    (hour[weather_columns] > 1)
).sum()

weather_range_validation = pd.DataFrame({
    'variable': weather_columns,
    'day_out_of_range': [
        day_weather_out_of_range[column]
        for column in weather_columns
    ],
    'hour_out_of_range': [
        hour_weather_out_of_range[column]
        for column in weather_columns
    ]
})

display(weather_range_validation)

,variable,day_out_of_range,hour_out_of_range
0,temp,0,0
1,atemp,0,0
2,hum,0,0
3,windspeed,0,0


### Weather range validation result
All normalized weather variables in both datasets fall within the expected range of 0 to 1.
No out-of-range weather values were identified.

## 11. Weather Boundary Value Review

This section counts exact zero values in the normalized weather variables and reviews potentially unusual boundary observations.

In [42]:
day_weather_zero_counts = (
    day[weather_columns] == 0
).sum()

hour_weather_zero_counts = (
    hour[weather_columns] == 0
).sum()

weather_zero_summary = pd.DataFrame({
    'variable': weather_columns,
    'day_zero_count': [
        day_weather_zero_counts[column]
        for column in weather_columns
    ],
    'hour_zero_count': [
        hour_weather_zero_counts[column]
        for column in weather_columns
    ]
})

display(weather_zero_summary)

,variable,day_zero_count,hour_zero_count
0,temp,0,0
1,atemp,0,2
2,hum,1,22
3,windspeed,0,2180


In [43]:
print('daily records with zero humidity:')
display(
    day.loc[
        day['hum'] == 0,
        ['dteday', 'season', 'weathersit', 'temp', 'hum', 'windspeed', 'cnt']
    ]
)

print('hourly records with zero humidity:')
display(
    hour.loc[
        hour['hum'] == 0,
        ['dteday', 'hr', 'season', 'weathersit', 'temp', 'hum', 'windspeed', 'cnt']
    ]
)

print('hourly records with zero apparent temperature:')
display(
    hour.loc[
        hour['atemp'] == 0,
        ['dteday', 'hr', 'temp', 'atemp', 'hum', 'windspeed', 'cnt']
    ]
)

daily records with zero humidity:


,dteday,season,weathersit,temp,hum,windspeed,cnt
68,2011-03-10,1,3,0.389091,0.0,0.261877,623


hourly records with zero humidity:


,dteday,hr,season,weathersit,temp,hum,windspeed,cnt
1551,2011-03-10,0,1,3,0.34,0.0,0.2537,3
1552,2011-03-10,1,1,3,0.34,0.0,0.2537,2
1553,2011-03-10,2,1,3,0.34,0.0,0.2537,1
1554,2011-03-10,5,1,3,0.36,0.0,0.1940,3
1555,2011-03-10,6,1,3,0.36,0.0,0.3284,12
1556,2011-03-10,7,1,3,0.38,0.0,0.2239,37
1557,2011-03-10,8,1,3,0.38,0.0,0.2836,44
1558,2011-03-10,9,1,3,0.40,0.0,0.2239,24
1559,2011-03-10,10,1,3,0.40,0.0,0.1642,17
1560,2011-03-10,11,1,3,0.40,0.0,0.2537,11


hourly records with zero apparent temperature:


,dteday,hr,temp,atemp,hum,windspeed,cnt
480,2011-01-22,1,0.04,0.0,0.41,0.3881,12
486,2011-01-22,8,0.02,0.0,0.44,0.3284,27


### Weather boundary review conclusion

The weather boundary-value review identified the following observations:

- Zero humidity is concentrated on March 10, 2011, with one daily record and 22 hourly records. This unusual pattern has been recorded as issue `dq-02`.
- Two zero apparent-temperature values occur on January 22, 2011. Their corresponding temperature values are also close to zero, so they are considered plausible boundary observations.
- Zero wind-speed values remain within the valid normalized range and are not treated as missing values.

All original values are retained without modification during the raw data audit.

In [44]:
new_issue = pd.DataFrame({
    'issue_id': ['dq-02'],
    'dataset': ['day and hour'],
    'issue': ['zero humidity values'],
    'description': [
        'Humidity is recorded as zero on 2011-03-10 in one daily record and 22 hourly records.'
    ],
    'severity': ['medium'],
    'decision': [
        'retain the original values and evaluate their effect during later analysis'
    ]
})

if 'dq-02' not in issue_register['issue_id'].values:
    issue_register = pd.concat(
        [issue_register, new_issue],
        ignore_index=True
    )

display(issue_register)

,issue_id,dataset,issue,description,severity,decision
0,dq-01,hour,missing date-hour records,165 expected date-hour combinations are absent...,medium,retain the raw data and evaluate the effect be...
1,dq-02,day and hour,zero humidity values,Humidity is recorded as zero on 2011-03-10 in ...,medium,retain the original values and evaluate their ...


## 12. Identifier Sequence Validation

This section checks whether the `instant` identifier starts at 1, follows a continuous sequence, and ends at the total number of records in each dataset.

In [45]:
day_expected_instant = pd.Series(
    range(1, len(day) + 1),
    index=day.index
)

hour_expected_instant = pd.Series(
    range(1, len(hour) + 1),
    index=hour.index
)

instant_validation = pd.DataFrame({
    'dataset': ['day', 'hour'],

    'minimum_instant': [
        day['instant'].min(),
        hour['instant'].min()
    ],

    'maximum_instant': [
        day['instant'].max(),
        hour['instant'].max()
    ],

    'expected_maximum': [
        len(day),
        len(hour)
    ],

    'sequence_mismatch': [
        (day['instant'] != day_expected_instant).sum(),
        (hour['instant'] != hour_expected_instant).sum()
    ]
})

display(instant_validation)

,dataset,minimum_instant,maximum_instant,expected_maximum,sequence_mismatch
0,day,1,731,731,0
1,hour,1,17379,17379,0


### Identifier sequence validation result

The `instant` identifier starts at 1, follows a continuous sequence, and ends at the total number of records in both datasets.

No missing, duplicated, or out-of-sequence identifiers were identified.

## 13. Cross-Dataset Calendar Consistency

This section checks whether the calendar-related variables recorded in the hourly dataset are consistent with the corresponding daily records.

In [46]:
calendar_columns = [
    'season',
    'yr',
    'mnth',
    'holiday',
    'weekday',
    'workingday'
]

hour_calendar_unique_counts = (
    hour.groupby('dteday')[calendar_columns]
    .nunique()
)

hour_calendar_inconsistencies = (
    hour_calendar_unique_counts > 1
).sum()

display(
    hour_calendar_inconsistencies
    .rename('inconsistent_dates')
    .to_frame()
)

,inconsistent_dates
season,0
yr,0
mnth,0
holiday,0
weekday,0
workingday,0


In [47]:
hour_calendar_daily = (
    hour.groupby('dteday', as_index=False)[calendar_columns]
    .first()
)

display(hour_calendar_daily.head())

,dteday,season,yr,mnth,holiday,weekday,workingday
0,2011-01-01,1,0,1,0,6,0
1,2011-01-02,1,0,1,0,0,0
2,2011-01-03,1,0,1,0,1,1
3,2011-01-04,1,0,1,0,2,1
4,2011-01-05,1,0,1,0,3,1


In [48]:
day_calendar = day[
    ['dteday'] + calendar_columns
].copy()

In [49]:
calendar_cross_check = day_calendar.merge(
    hour_calendar_daily,
    on='dteday',
    how='left',
    suffixes=('_day', '_hour')
)

display(calendar_cross_check.head())

,dteday,season_day,yr_day,mnth_day,holiday_day,weekday_day,workingday_day,season_hour,yr_hour,mnth_hour,holiday_hour,weekday_hour,workingday_hour
0,2011-01-01,1,0,1,0,6,0,1,0,1,0,6,0
1,2011-01-02,1,0,1,0,0,0,1,0,1,0,0,0
2,2011-01-03,1,0,1,0,1,1,1,0,1,0,1,1
3,2011-01-04,1,0,1,0,2,1,1,0,1,0,2,1
4,2011-01-05,1,0,1,0,3,1,1,0,1,0,3,1


In [50]:
cross_dataset_calendar_summary = pd.DataFrame({
    'variable': calendar_columns,

    'mismatch_dates': [
        (
            calendar_cross_check[f'{column}_day'] !=
            calendar_cross_check[f'{column}_hour']
        ).sum()
        for column in calendar_columns
    ]
})

display(cross_dataset_calendar_summary)

,variable,mismatch_dates
0,season,0
1,yr,0
2,mnth,0
3,holiday,0
4,weekday,0
5,workingday,0


### Cross-dataset calendar consistency result

The calendar-related variables are internally consistent within the hourly dataset and fully match the corresponding daily records.

No cross-dataset calendar inconsistencies were identified.

## 14. Daily and Hourly Weather Consistency

This section compares the daily weather values with the mean hourly weather values for each date.

In [51]:
weather_columns = [
    'temp',
    'atemp',
    'hum',
    'windspeed'
]

hourly_daily_weather = (
    hour.groupby('dteday', as_index=False)[weather_columns]
    .mean()
    .rename(columns={
        'temp': 'hour_temp_mean',
        'atemp': 'hour_atemp_mean',
        'hum': 'hour_hum_mean',
        'windspeed': 'hour_windspeed_mean'
    })
)

display(hourly_daily_weather.head())

,dteday,hour_temp_mean,hour_atemp_mean,hour_hum_mean,hour_windspeed_mean
0,2011-01-01,0.344167,0.363625,0.805833,0.160446
1,2011-01-02,0.363478,0.353739,0.696087,0.248539
2,2011-01-03,0.196364,0.189405,0.437273,0.248309
3,2011-01-04,0.200000,0.212122,0.590435,0.160296
4,2011-01-05,0.226957,0.229270,0.436957,0.186900


In [52]:
daily_weather = day[
    ['dteday'] + weather_columns
].rename(columns={
    'temp': 'day_temp',
    'atemp': 'day_atemp',
    'hum': 'day_hum',
    'windspeed': 'day_windspeed'
})

weather_cross_check = daily_weather.merge(
    hourly_daily_weather,
    on='dteday',
    how='left'
)

display(weather_cross_check.head())

,dteday,day_temp,day_atemp,day_hum,day_windspeed,hour_temp_mean,hour_atemp_mean,hour_hum_mean,hour_windspeed_mean
0,2011-01-01,0.344167,0.363625,0.805833,0.160446,0.344167,0.363625,0.805833,0.160446
1,2011-01-02,0.363478,0.353739,0.696087,0.248539,0.363478,0.353739,0.696087,0.248539
2,2011-01-03,0.196364,0.189405,0.437273,0.248309,0.196364,0.189405,0.437273,0.248309
3,2011-01-04,0.200000,0.212122,0.590435,0.160296,0.200000,0.212122,0.590435,0.160296
4,2011-01-05,0.226957,0.229270,0.436957,0.186900,0.226957,0.229270,0.436957,0.186900


In [53]:
weather_difference_summary = pd.DataFrame({
    'variable': weather_columns,

    'maximum_absolute_difference': [
        (
            weather_cross_check['day_temp'] -
            weather_cross_check['hour_temp_mean']
        ).abs().max(),

        (
            weather_cross_check['day_atemp'] -
            weather_cross_check['hour_atemp_mean']
        ).abs().max(),

        (
            weather_cross_check['day_hum'] -
            weather_cross_check['hour_hum_mean']
        ).abs().max(),

        (
            weather_cross_check['day_windspeed'] -
            weather_cross_check['hour_windspeed_mean']
        ).abs().max()
    ]
})

display(weather_difference_summary)

,variable,maximum_absolute_difference
0,temp,4.782609e-07
1,atemp,5.000000e-07
2,hum,5.217391e-07
3,windspeed,5.000000e-07


### Daily-hourly weather consistency result

The daily temperature, apparent temperature, humidity, and wind-speed values closely match the corresponding mean hourly values.

The maximum absolute differences are below 0.000001 and are attributable to numerical rounding. No meaningful weather inconsistencies were identified between the two datasets.

## 15. Data Quality Summary

This section summarizes the main data-quality checks and their results.

In [54]:
data_quality_summary = pd.DataFrame({
    'check_id': [
        'qc-01',
        'qc-02',
        'qc-03',
        'qc-04',
        'qc-05',
        'qc-06',
        'qc-07',
        'qc-08',
        'qc-09',
        'qc-10',
        'qc-11',
        'qc-12',
        'qc-13'
    ],

    'quality_check': [
        'cell-level missing values',
        'duplicate records',
        'categorical value validity',
        'rental count integrity',
        'date validity and coverage',
        'calendar consistency',
        'identifier sequence',
        'hourly record completeness',
        'daily-hourly demand reconciliation',
        'weather value ranges',
        'weather boundary values',
        'cross-dataset calendar consistency',
        'daily-hourly weather consistency'
    ],

    'dataset': [
        'day and hour',
        'day and hour',
        'day and hour',
        'day and hour',
        'day and hour',
        'day and hour',
        'day and hour',
        'hour',
        'day and hour',
        'day and hour',
        'day and hour',
        'day and hour',
        'day and hour'
    ],

    'status': [
        'pass',
        'pass',
        'pass',
        'pass',
        'pass',
        'pass',
        'pass',
        'issue identified',
        'pass',
        'pass',
        'issue identified',
        'pass',
        'pass'
    ],

    'finding': [
        'No cell-level missing values were found.',
        'No duplicate rows or duplicate key values were found.',
        'All categorical variables contain expected codes.',
        'No negative counts were found and cnt equals casual plus registered.',
        'All dates are valid and both datasets cover 2011-01-01 to 2012-12-31.',
        'Year, month, weekday, holiday, and workingday values are consistent.',
        'The instant identifier is unique, complete, and sequential.',
        '165 expected date-hour records are absent across 76 dates.',
        'Aggregated hourly rental counts exactly match the daily totals.',
        'All normalized weather values fall within the range of 0 to 1.',
        'Zero humidity occurs on 2011-03-10 in one daily and 22 hourly records.',
        'Calendar variables match between the daily and hourly datasets.',
        'Daily weather values match mean hourly values within rounding tolerance.'
    ]
})

display(data_quality_summary)

,check_id,quality_check,dataset,status,finding
0,qc-01,cell-level missing values,day and hour,pass,No cell-level missing values were found.
1,qc-02,duplicate records,day and hour,pass,No duplicate rows or duplicate key values were...
2,qc-03,categorical value validity,day and hour,pass,All categorical variables contain expected codes.
3,qc-04,rental count integrity,day and hour,pass,No negative counts were found and cnt equals c...
4,qc-05,date validity and coverage,day and hour,pass,All dates are valid and both datasets cover 20...
5,qc-06,calendar consistency,day and hour,pass,"Year, month, weekday, holiday, and workingday ..."
6,qc-07,identifier sequence,day and hour,pass,"The instant identifier is unique, complete, an..."
7,qc-08,hourly record completeness,hour,issue identified,165 expected date-hour records are absent acro...
8,qc-09,daily-hourly demand reconciliation,day and hour,pass,Aggregated hourly rental counts exactly match ...
9,qc-10,weather value ranges,day and hour,pass,All normalized weather values fall within the ...


In [55]:
quality_status_summary = (
    data_quality_summary['status']
    .value_counts()
    .rename_axis('status')
    .reset_index(name='check_count')
)

display(quality_status_summary)

,status,check_count
0,pass,11
1,issue identified,2


### Data quality summary conclusion

The raw data audit completed 13 major quality checks. Eleven checks passed without identifying material inconsistencies.

Two issues were identified:

1. The hourly dataset is structurally incomplete, with 165 absent date-hour records across 76 dates.
2. Zero humidity is recorded on March 10, 2011, in one daily record and 22 hourly records.

The identified issues have been documented in the issue register. No raw values or records were modified during the audit.

## 16. Final Issue Register and Data Readiness Decision

This section presents the final issue register and determines whether the datasets are ready for cleaning, transformation, and analysis.

In [56]:
issue_register = (
    issue_register
    .sort_values('issue_id')
    .reset_index(drop=True)
)

display(issue_register)

,issue_id,dataset,issue,description,severity,decision
0,dq-01,hour,missing date-hour records,165 expected date-hour combinations are absent...,medium,retain the raw data and evaluate the effect be...
1,dq-02,day and hour,zero humidity values,Humidity is recorded as zero on 2011-03-10 in ...,medium,retain the original values and evaluate their ...


In [57]:
data_readiness_summary = pd.DataFrame({
    'dataset': [
        'day',
        'hour'
    ],

    'readiness_status': [
        'ready with minor review',
        'ready with conditions'
    ],

    'reason': [
        (
            'The daily dataset passed the main integrity checks, '
            'but the zero-humidity observation on 2011-03-10 '
            'must remain documented.'
        ),
        (
            'The hourly dataset passed the main integrity checks, '
            'but 165 date-hour records are absent across 76 dates.'
        )
    ],

    'required_action': [
        (
            'Retain the original zero-humidity value and evaluate '
            'its influence during later analysis.'
        ),
        (
            'Do not interpret absent hourly records as zero demand. '
            'Document the gaps and evaluate their effect before '
            'hour-level modeling.'
        )
    ]
})

display(data_readiness_summary)

,dataset,readiness_status,reason,required_action
0,day,ready with minor review,The daily dataset passed the main integrity ch...,Retain the original zero-humidity value and ev...
1,hour,ready with conditions,The hourly dataset passed the main integrity c...,Do not interpret absent hourly records as zero...


### Final data readiness decision

The raw daily and hourly datasets are suitable for the next stage of the project, subject to the documented data-quality conditions.

The daily dataset is ready for cleaning and analysis with minor review of the zero-humidity observation on March 10, 2011.

The hourly dataset is ready with conditions. The 165 absent date-hour records must not be interpreted as zero-demand observations, and their effect should be evaluated before hour-level analysis or modeling.

No raw records or values were modified during this audit.

## 17. Export Audit Outputs

This section exports the final data-quality summary, issue register, readiness decision, and hourly-gap diagnostics for documentation and later project stages.

In [58]:
tables_output_dir = project_root / '04_outputs' / 'tables'
diagnostics_output_dir = project_root / '04_outputs' / 'diagnostics'

tables_output_dir.mkdir(parents=True, exist_ok=True)
diagnostics_output_dir.mkdir(parents=True, exist_ok=True)

print('tables output folder:', tables_output_dir)
print('diagnostics output folder:', diagnostics_output_dir)

tables output folder: d:\کلاس دیتا\پاسخ تمارین\مصور سازی\fatemehkamrani_bike_sharing_demand_analysis\04_outputs\tables
diagnostics output folder: d:\کلاس دیتا\پاسخ تمارین\مصور سازی\fatemehkamrani_bike_sharing_demand_analysis\04_outputs\diagnostics


In [59]:
data_quality_summary.to_csv(
    tables_output_dir / 'data_quality_summary.csv',
    index=False
)

issue_register.to_csv(
    tables_output_dir / 'issue_register.csv',
    index=False
)

data_readiness_summary.to_csv(
    tables_output_dir / 'data_readiness_summary.csv',
    index=False
)

print('summary tables were exported successfully.')

summary tables were exported successfully.


In [60]:
incomplete_hourly_dates.to_csv(
    diagnostics_output_dir / 'incomplete_hourly_dates.csv',
    index=False
)

missing_hours_by_date.to_csv(
    diagnostics_output_dir / 'missing_hours_by_date.csv',
    index=False
)

print('hourly gap diagnostics were exported successfully.')

hourly gap diagnostics were exported successfully.


In [61]:
audit_output_files = [
    tables_output_dir / 'data_quality_summary.csv',
    tables_output_dir / 'issue_register.csv',
    tables_output_dir / 'data_readiness_summary.csv',
    diagnostics_output_dir / 'incomplete_hourly_dates.csv',
    diagnostics_output_dir / 'missing_hours_by_date.csv'
]

for file in audit_output_files:
    print(file.name, ':', file.exists())

data_quality_summary.csv : True
issue_register.csv : True
data_readiness_summary.csv : True
incomplete_hourly_dates.csv : True
missing_hours_by_date.csv : True


### Audit completion

The raw data audit has been completed successfully.

The final data-quality summary, issue register, readiness decision, and hourly-gap diagnostics were exported to the project output folders.

The raw `day.csv` and `hour.csv` files remained unchanged throughout the audit.